`types.path`: Do not allow lists of strings

roberth opened the pull request:
###### Description of changes

Explains https://github.com/NixOS/nixpkgs/issues/208044

`types.path` used to allow lists of strings, because they are coercible to strings when `more = true` in Nix. However, lists of stringlike values are rarely intended to be actual paths, because spaces in paths are quite rare, and better represented by actual spaces. Therefore I believe the old behavior to be a bug, and expect this change to fix more things than it breaks.

This PR further improves `lib.strings` by renaming `lib.strings.isCoercibleToString` and deprecating it after 23.05, in order to make the `more = true | false` setting explicit.

`listOf path` will behave as before, unless individual list items were in turn lists of paths.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] ~Tested, as applicable~ compared to the Nix C++ code
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

roberth committed:
lib.strings: Add isSimpleCoercibleToString

roberth committed:
lib.strings.isStorePath: Use isSimpleCoercibleToString

Expecting no change in behavior.

roberth committed:
lib.types.anything: Use isSimpleCoercibleToString

Expecting no change in behavior.

roberth committed:
lib.strings: Rename isCoercibleToString -> isMoreCoercibleToString

roberth committed:
treewide: isCoercibleToString -> isMoreCoercibleToString

No change in behavior.

roberth committed:
lib.strings.toShellVar: Use isSimpleCoercibleString

Expecting no change in behavior.

roberth committed:
lib.types.path: Do not allow lists of strings

roberth committed:
nixos/self-deploy: Cleanup after types.path is not allowed to be a list anymore

figsoda commented:
I don't like the naming of the two functions, I think the original `isCoercibleToString` is clear in what it means: something that can be coerced to a string with `toString`

I would prefer something like `isStringLike` over `isSimpleCoercibleToString`, or just inlining the function in `types.path`

I do agree that `types.path` shouldn't accept a list of strings

roberth commented:
> naming of the two functions

I've tried to stay close to both the existing name and the Nix (C++) code. "Stringlike" is not an existing concept in Nix or Nixpkgs.

> I think the original `isCoercibleToString` is clear in what it means

It appears clear, but people have used it wrong because the name is not precise. Ostensibly they've expected it to work like `coerceToString(more = false)`, but we can't just change the meaning of the function.
Arguably they were reasonable to make that assumption. In Nix, the default is `more = false`, and it is about coercions, not the `toString` function. These coercions are used for string concatenation, interpolation, functions like `baseNameOf`, etc, so people expect this.

ncfavier reviewed (commented):
This is much needed.

Regarding naming: I agree that `isCoercibleToString` is ambiguous and should go. I think the chosen names are fine, but I think we could probably be bold and coin "string-like" for the simple coercion behaviour.

Also the new functions should probably be exposed in `lib/default.nix`.

roberth committed:
lib.strings: isSimpleCoercibleString -> isStringLike

roberth committed:
lib.strings: isMoreCoercibleString -> isConvertibleWithToString

Yes, this function name is inconveniently long, but it is important
for the name to explicitly reference the function and not be mistaken
for the implicit string conversions, which only happen for a smaller
set of values.

roberth committed:
lib.strings.isConvertibleWithToString: Refactor to reuse isStringLike

roberth committed:
lib: Add isPath

Available since Nix 2.3, which is the Nixpkgs minimum version.
Thanks zimbatm!

roberth committed:
lib.isStringLike: Remove use of list

In the current implementation of Nix, this list would be allocated
over and over. Iirc pennae tried to optimize static list allocation,
but gained no significant performance improvement.

roberth commented:
> I would prefer something like `isStringLike`

> > "Stringlike" is not an existing concept in Nix or Nixpkgs.

> I think we could probably be bold and coin "string-like" for the simple coercion behaviour.

You've convinced me. `isStringLike` it is.

I've changed the names and I've added two more commits at the end hoping for a performance improvement. I'm waiting for the ofborg eval performance report for that.

ncfavier commented on lib/strings.nix:
```diff
@@ -798,10 +799,31 @@ rec {
+  /* Check whether a value can be coerced to a string.
+     The value must be a string, path, or attribute set.
+
+     StringLike values can be used without explicit conversion in most functions
+     that expect a string.
```
```suggestion
     String-like values can be used without explicit conversion in most places
     that expect a string.
```

Importantly, in interpolations.

I've actually had in my TODO list for a while to add a `types.stringLike` meant to replace the `either str path` (or, worse, `either str package`) idiom e.g. in [`system.build.installBootLoader`](https://github.com/NixOS/nixpkgs/blob/83d75023879180ffe21169216ad86573a153a303/nixos/modules/system/activation/top-level.nix#L188), [`console.font`](https://github.com/NixOS/nixpkgs/blob/83d75023879180ffe21169216ad86573a153a303/nixos/modules/config/console.nix#L48), etc. Essentially options whose type is "anything Nix accepts in an interpolation".

roberth replied:
Applied.



> [`console.font`](https://github.com/NixOS/nixpkgs/blob/83d75023879180ffe21169216ad86573a153a303/nixos/modules/config/console.nix#L48),

I think the current type for this one is actually nice, because the two types have quite different interpretations. It adds a bit of meaning compared to `stringLike`, even if that would be sufficient.

ncfavier commented:
I think we should have at least `lib.isStringLike` in the top level since we have `lib.isString`.

roberth committed:
lib: Add isStringLike

roberth commented:
> I think we should have at least `lib.isStringLike` in the top level since we have `lib.isString`.

Done.

I don't expect a lot of use of the other function.
